genesyscloud/journey_segment/resource_genesyscloud_journey_segment_utils.go

```diff
@@ -0,0 +1,260 @@
+	journey := resourcedata.BuildSdkListFirstElement(journeySegment, "journey", buildSdkPatchJourney, false)
+
+	sdkPatchSegment := platformclientv2.Patchsegment{}
+	sdkPatchSegment.SetField("IsActive", &isActive)
+	sdkPatchSegment.SetField("DisplayName", &displayName)
+	sdkPatchSegment.SetField("Description", description)
+	sdkPatchSegment.SetField("Color", &color)
+	sdkPatchSegment.SetField("ShouldDisplayToAgent", shouldDisplayToAgent)
+	sdkPatchSegment.SetField("Context", sdkContext)
+	sdkPatchSegment.SetField("Journey", journey)
+	return &sdkPatchSegment
+}
+
+func flattenContext(context *platformclientv2.Context) map[string]interface{} {
+	if len(*context.Patterns) == 0 {
+		return nil
+	}
+	contextMap := make(map[string]interface{})
+	contextMap["patterns"] = *lists.FlattenList(context.Patterns, flattenContextPattern)
+	return contextMap
+}
+
+func buildSdkRequestContext(context map[string]interface{}) *platformclientv2.Requestcontext {
+	patterns := &[]platformclientv2.Requestcontextpattern{}
+	if context != nil {
+		patterns = stringmap.BuildSdkList(context, "patterns", buildSdkRequestContextPattern)
+	}
+	return &platformclientv2.Requestcontext{
+		Patterns: patterns,
+	}
+}
+
+func buildSdkPatchContext(context map[string]interface{}) *platformclientv2.Patchcontext {
+	patterns := &[]platformclientv2.Patchcontextpattern{}
+	if context != nil {
+		patterns = stringmap.BuildSdkList(context, "patterns", buildSdkPatchContextPattern)
+	}
+	return &platformclientv2.Patchcontext{
+		Patterns: patterns,
+	}
+}
+
+func flattenContextPattern(contextPattern *platformclientv2.Contextpattern) map[string]interface{} {
+	contextPatternMap := make(map[string]interface{})
+	contextPatternMap["criteria"] = *lists.FlattenList(contextPattern.Criteria, flattenEntityTypeCriteria)
+	return contextPatternMap
+}
+
+func buildSdkRequestContextPattern(contextPattern map[string]interface{}) *platformclientv2.Requestcontextpattern {
+	return &platformclientv2.Requestcontextpattern{
+		Criteria: stringmap.BuildSdkList(contextPattern, "criteria", buildSdkRequestEntityTypeCriteria),
+	}
+}
+
+func buildSdkPatchContextPattern(contextPattern map[string]interface{}) *platformclientv2.Patchcontextpattern {
+	return &platformclientv2.Patchcontextpattern{
+		Criteria: stringmap.BuildSdkList(contextPattern, "criteria", buildSdkPatchEntityTypeCriteria),
+	}
+}
+
+func flattenEntityTypeCriteria(entityTypeCriteria *platformclientv2.Entitytypecriteria) map[string]interface{} {
+	entityTypeCriteriaMap := make(map[string]interface{})
+	entityTypeCriteriaMap["key"] = *entityTypeCriteria.Key
+	entityTypeCriteriaMap["values"] = lists.StringListToSet(*entityTypeCriteria.Values)
+	entityTypeCriteriaMap["should_ignore_case"] = *entityTypeCriteria.ShouldIgnoreCase
+	entityTypeCriteriaMap["operator"] = *entityTypeCriteria.Operator
+	entityTypeCriteriaMap["entity_type"] = *entityTypeCriteria.EntityType
```

Comment:
This is a good place to use `resourcedata.SetMapValueIfNotNil`. Seems like this function is a prime suspect for future stack traces